pkg/tests/ossm/smoke_test.go

```diff
@@ -15,55 +15,40 @@
 package ossm
 
 import (
-	"strings"
+	"net/http"
 	"testing"
-	"time"
 
 	"github.com/maistra/maistra-test-tool/pkg/app"
-	. "github.com/maistra/maistra-test-tool/pkg/examples"
-	. "github.com/maistra/maistra-test-tool/pkg/util"
+	"github.com/maistra/maistra-test-tool/pkg/util/check/require"
+	"github.com/maistra/maistra-test-tool/pkg/util/curl"
 	. "github.com/maistra/maistra-test-tool/pkg/util/log"
-	"github.com/maistra/maistra-test-tool/pkg/util/shell"
+	"github.com/maistra/maistra-test-tool/pkg/util/oc"
+	"github.com/maistra/maistra-test-tool/pkg/util/retry"
 	. "github.com/maistra/maistra-test-tool/pkg/util/test"
 )
 
-func cleanupBookinfo() {
-	Log.Info("Cleanup")
-	app := Bookinfo{"bookinfo"}
-	app.Uninstall()
-	time.Sleep(time.Duration(30) * time.Second)
-}
-
 func TestBookinfo(t *testing.T) {
 	NewTest(t).Id("A2").Groups(ARM, Full, Smoke, InterOp).Run(func(t TestHelper) {
-		Log.Info("Test Bookinfo Installation")
-		app.InstallAndWaitReady(t, app.Bookinfo("bookinfo"))
 
-		Log.Info("Check pods running 2/2 ready")
-		msg, _ := Shell(`oc get pods -n bookinfo`)
-		if strings.Contains(msg, "2/2") {
-			Log.Info("Success. proxy container is running.")
-		} else {
-			t.Error("Error. proxy container is not running.")
-		}
+		ns := "bookinfo"
+
+		t.Cleanup(func() {
+			oc.RecreateNamespace(t, ns)
+		})
 
-		Log.Info("Check istiod pod is ready and print istiod logs")
-		mesg, _ := Shell(`oc get pods -n istio-system | grep istiod`)
-		if strings.Contains(mesg, "1/1") {
-			Log.Info("Success. istiod pod is running with below logs:")
-			shell.Executef(t, `oc logs -n %s -l app=istiod | grep info`, meshNamespace)
-		} else {
-			t.Error("Error. istiod pod is not running.")
-		}
+		app.InstallAndWaitReady(t, app.Bookinfo(ns))
 
 		Log.Info("Check if bookinfo productpage is running")
-		GATEWAY_URL, _ := Shell(`oc -n %s get route istio-ingressgateway -o jsonpath='{.spec.host}'`, meshNamespace)
-		mes, _ := Shell(`curl -o /dev/null -s -w "%%{http_code}\n" http://%s/productpage`, GATEWAY_URL)
-		if strings.Contains(mes, "200") {
-			Log.Info("Success. bookinfo productpage is running")
-		} else {
-			t.Error("Error. bookinfo productpage is not running.")
-			Log.Error("Error. bookinfo productpage is not running.")
-		}
+
+		productpageURL := app.BookinfoProductPageURL(t, meshNamespace)
+
+		t.LogStep("check if productpage shows 'error fetching product reviews' due to delay injection")
+		retry.UntilSuccess(t, func(t TestHelper) {
+			for i := 0; i < 5; i++ {
+				curl.Request(t,
+					productpageURL, nil,
+					require.ResponseStatus(http.StatusOK))
```

Comment:
Instead of checking just the status, you could use `require.ResponseMatchesFile` to check that we're getting back the actual product page response.

Reply:
Also, we could improve this test by checking if the response was served by envoy. This way, the test would confirm that everything, including istio-cni, is working as it should. 

Let me explain: if you only check if all the pods' status is `2/2` (you will do this above), this only means that the sidecar is injected, but doesn't necessarily mean that everything is fine. If istio-cni hasn't done its job, the requests to productpage won't go through the proxy. That's why this test should verify that requests are going through the proxy. You can do that by checking the response headers. 

But, if you use productpageURL, which goes through the ingress gateway, you will always get that header, so you have to go directly to the service like this:
```
$ kubectl -n bookinfo run -it --restart=Never --rm curl --image curlimages/curl -- curl -I http://productpage:9080
HTTP/1.1 200 OK
content-type: text/html; charset=utf-8
content-length: 1683
server: istio-envoy
date: Tue, 04 Apr 2023 06:43:54 GMT
x-envoy-upstream-service-time: 2
x-envoy-decorator-operation: productpage.bookinfo.svc.cluster.local:9080/*
```

You can then assert that the output contains either the string `server: istio-envoy` or `x-envoy`.